### ISSUE TITLE
with_read_replica for more GET endpoints

### ISSUE THREAD

**Issue by ryan-augustinsky** [mastodon/mastodon#28649]
### Pitch

After poking around, I've noticed there are many high-traffic GET endpoints that could use with_read_replica.

This is a feature that could greatly benefit from a little extra work.

### Motivation

I'm testing out distributing web servers to New York City, San Francisco, and Frankfurt with a geo load balancer.

To reduce postgres query latency, there will be read replicas in each region at the same data centers.

Requests that are not GET are sent directly to Frankfurt where the master database is located.

GET requests that hit the regional web server should run low-latency queries to the regional read-only replica database most of the time. 

When there are many database queries called from one endpoint request, each query to the master database could add more than 100ms to the final cumulative request time.

Note: I realize redis can't be distributed like this right now, this will be a long-term project of mine.

**Comment by renchap** [comment]
This is not a supported setup, and it will need major architecture changes to support it. Web servers and workers should be hosted locally with the database & redis (ie, <1ms latency) if you want correct performance.

**Comment by ryan-augustinsky** [comment]
> This is not a supported setup, and it will need major architecture changes to support it. Web servers and workers should be hosted locally with the database & redis (ie, <1ms latency) if you want correct performance.

I mostly have this already, but minus redis.
Right now I'm experimenting with wrapping all API requests in with_read_replica.

  around_action :wrap_with_read_replica, only: [:show, :index]
  def wrap_with_read_replica
    with_read_replica do 
      yield
    end
  end

These hit the replica db in the same regional data center.
All non-get requests hit a Frankfurt web server where the write database also lives.

This is all beside the point though. If there are more endpoints that can be wrapped with_read_replica, why not wrap them? :smile:

**Comment by mjankowski** [comment]
> This is all beside the point though. If there are more endpoints that can be wrapped with_read_replica, why not wrap them? 😄

Curious about this as well.

Seems like the original question was sidetracked a little due to the geography component.

Assuming every single server is sitting on the same network shelf right next to each other -- are there still reasons to not expand the read replica configuration (doing a first pass in one app area for what would be confident well-known non-write actions).

**Comment by renchap** [comment]
> Assuming every single server is sitting on the same network shelf right next to each other -- are there still reasons to not expand the read replica configuration (doing a first pass in one app area for what would be confident well-known non-write actions).

No, and historically most `SELECT` were sent to the replica before we switched to AR's built-in replica support due to the way the gem we used worked.

But it can create subttle issues if the replica is lagging behind, for example if something is created on the primary DB, but is not yet on the replica, and a request immediately made after the first one hits the replica where the record does not exists.

So we went with a conservative approach, where we send all queries to the primary by default. For most requests, they are not called veyr often, so it does not have an impact, and this is the safer. Then we looked at our production query statistics, and moved the most resource intensive queries (based on total time, so number of calls * average time) to the replica where it made sense because the situation above could not happen.

**Comment by mjankowski** [comment]
Thanks for summary.

> But it can create subttle issues if the replica is lagging behind, for example if something is created on the primary DB, but is not yet on the replica, and a request immediately made after the first one hits the replica where the record does not exists.

For that issue specifically, maybe future research - https://guides.rubyonrails.org/active_record_multiple_databases.html#activating-automatic-role-switching - specifically the "read your own write" with configurable time window portion.

> [...] and moved the most resource intensive queries (based on total time, so number of calls * average time) to the replica where it made sense because the situation above could not happen.

Makes total sense as first pass.

LMK if any investigation into that role switching feature would be useful.
